src/Data/Multimap/Generic.hs

```diff
@@ -224,6 +248,15 @@ mapGroups
   => (Group k1 (c1 v1) -> Group k2 (c2 v2)) -> Multimap c1 k1 v1 -> Multimap c2 k2 v2
 mapGroups f = fromGroupList . fmap f . toGroupList
 
+-- | @mapGroupKeysWith c f s@ is the multimap obtained by applying @f@ to each key of @s@.
+--
+-- The size of the result may be smaller if @f@ two or more distinct keys to the same new key. In this case the
+-- associated values will be combined using @c@. The value at the greater of the two original keys is used as the first
+-- argument to @c@.
+mapGroupKeysWith :: (Collection c, Ord k2)
+                 => (c a -> c a -> c a) -> (k1 -> k2) -> Multimap c k1 a -> Multimap c k2 a
```

Comment:
Can't we use the existing `mapGroups` for this?

Reply:
Yes. I thought it would be less efficient, but now I checked and `Map.mapKeys` is also going through a list representation. I can remove it if you wish.